File: scripts/generate_changelog_header.py
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
SOURCE_PATH = ROOT / "CHANGELOG.md"
OUTPUT_PATH = ROOT / "headers" / "ChangelogData.h"


def escape(value):
    return value.replace("\\", "\\\\").replace('"', '\\"').replace("\r", "").replace("\n", "\\n")


def main():
    contents = SOURCE_PATH.read_text() if SOURCE_PATH.exists() else "# Changelog\n"
    OUTPUT_PATH.write_text(f'#define GONERINO_CHANGELOG @"{escape(contents)}"\n')


if __name__ == "__main__":
    main()

File: scripts/generate_changelog_header.go
package main

import (
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, source, _, _ := runtime.Caller(0)
	root := filepath.Dir(filepath.Dir(source))
	if override := os.Getenv("GONERINO_ROOT"); override != "" {
		root = override
	}
	sourcePath := filepath.Join(root, "CHANGELOG.md")
	outputPath := filepath.Join(root, "headers", "ChangelogData.h")
	contents, err := os.ReadFile(sourcePath)
	if err != nil {
		if !os.IsNotExist(err) {
			panic(err)
		}
		contents = []byte("# Changelog\n")
	}
	escaped := strings.NewReplacer(`\`, `\\`, `"`, `\"`, "\r", "", "\n", `\n`).Replace(string(contents))
	output := []byte(`#define GONERINO_CHANGELOG @"` + escaped + `"` + "\n")
	if err := os.WriteFile(outputPath, output, 0644); err != nil {
		panic(err)
	}
}
